### PR TITLE
Memory handling of 2 reads at the same address in the same clock cycle

### LINKED ISSUE #1560: `RCOMBBASE` is broken
Failing test:

```rs
#[test]
fn rcomb_base() {
    const X_PTR: u64 = 42;
    const Z_PTR: u64 = 142;
    const A_PTR: u64 = 1142;
    let asm_op = "begin rcomb_base end";

    let stack_inputs = [1,2,3,4,5,6,7,0,0,X_PTR, Z_PTR, A_PTR];
    let test = build_test!(asm_op, &stack_inputs);
    let _last_state = test.get_last_stack_state();
}
```

This test case also finds another bug in the memory chiplet [here](https://github.com/0xPolygonMiden/miden-vm/blob/766292d4efd9bcf663528631410a2d5361f3bb69/processor/src/chiplets/memory/mod.rs#L180-L182), where the fix is

```rs
let delta = if prev_ctx != ctx {
    (u32::from(ctx) - u32::from(prev_ctx)).into()
} else if prev_addr != addr {
    u64::from(addr - prev_addr)
} else if clk != prev_clk {
    clk - prev_clk - 1
} else {
    0
};
```

### PR DESCRIPTION
**Pull request by plafer**
Closes #1560

### CODE REVIEW

File: miden/tests/integration/operations/io_ops/mem_ops.rs [hunk 1]
```diff
@@ -228,3 +228,16 @@ fn read_after_write() {
     let test = build_op_test!("mem_storew.0 dropw mem_loadw.0", &[1, 2, 3, 4, 5, 6, 7, 8]);
     test.expect_stack(&[8, 7, 6, 5]);
 }
+
+// MISC
+// ================================================================================================
+
+/// Ensures that the processor handles 2 memory operations in the same context, at the same address,
+/// in the same clock cycle (which is what RCOMBBASE does when `stack[13] = stack[14] = 0`).
```
Comment (Al-Kindi-0) [inline miden/tests/integration/operations/io_ops/mem_ops.rs:236]: I am not sure that `RCOMBBASE`  reads from the same address at the same cycle, at least this is not its intended use

Comment (plafer) [reply miden/tests/integration/operations/io_ops/mem_ops.rs:236]: I agree that it's not the intended use; this is just a way to show the limitation in our processor. Whether it's with `RCOMBBASE` or any other (future) instruction.

e.g. if there's a bug at some point and we execute `RCOMBBASE` on `stack[13] = stack[14] = 0`, then the VM will crash. The intention of this PR is to fix this.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Comment by Al-Kindi-0** [comment]
I would personally not treat this at the constraint system unless there is a real use case or a good reason for allowing reading the same address at the same cycles in the same context.
Nevertheless, if we go with this approach, we would have to update the constraints to take into account the changes in this PR. For example the following constraints needs to be updated as the cycles need not anymore be monotonically increasing
montonicaly_increasing

There could be others so we should double check all constraints which rely on assumptions changes with this PR.

**Comment by bobbinth** [comment]
> we would have to update the constraints to take into account the changes in this PR. For example the following constraints needs to be updated as the cycles need not anymore be monotonically increasing

Ah - good point. I guess there are a few ways we can go:

1. Go through with this PR and fix the constraints (either in this or a follow-up PR).
2. Make it explicit that reading from the same address in the same cycle is not allowed (and have a proper error for this).
3. Tackle it as a part of a bigger memory refactoring we'll do in December. For example, maybe it would make sense to introduce a "multiplicity" column for reads. This way this constraint would stay the same. Though, it may introduce more complexities.

Overall, to keep track of the things we may want to address with memory subsystem refactoring:
- Fix the current soundness bug for single element writes.
- Enable support for basic read-only memory.
- Potentially enable support for element-addressable memory.
- Potentially enable multiple reads from the same address at the same cycle.
- Maybe something else as well.

**Comment by Al-Kindi-0** [comment]
> > we would have to update the constraints to take into account the changes in this PR. For example the following constraints needs to be updated as the cycles need not anymore be monotonically increasing
> 
> Ah - good point. I guess there are a few ways we can go:
> 
>     1. Go through with this PR and fix the constraints (either in this or a follow-up PR).
> 
>     2. Make it explicit that reading from the same address in the same cycle is not allowed (and have a proper error for this).
> 
>     3. Tackle it as a part of a bigger memory refactoring we'll do in December. For example, maybe it would make sense to introduce a "multiplicity" column for reads. This way this constraint would stay the same. Though, it may introduce more complexities.
> 

Makes sense!
Up to you guys, I would prefer option 2 at the moment but given the work in this PR option 1 is fine as well.


> 
> Overall, to keep track of the things we may want to address with memory subsystem refactoring:
> 
>     * Fix the current soundness bug for single element writes.
> 
>     * Enable support for basic read-only memory.
> 
>     * Potentially enable support for element-addressable memory.
> 
>     * Potentially enable multiple reads from the same address at the same cycle.
> 
>     * Maybe something else as well.

I think that's all

**Comment by plafer** [comment]
> I would personally not treat this at the constraint system unless there is a real use case or a good reason for allowing reading the same address at the same cycles in the same context.

Ah yes you're right! Re-reading this now, it does seem like the right approach is to simply return an error in the processor, instead of changing the constraints to allow it. We currently never need to read/write the same location twice, and so it doesn't make sense to allow it.

I will update the PR accordingly.

**Comment by plafer** [comment]
Implementation changed to return an error when a memory address is accessed twice in the same clock cycle.

Interestingly the `rcombine_main()` test was never incrementing the clock and so was reading & writing the same memory address in the same clock cycle (now fixed).

File: processor/src/operations/comb_ops.rs [hunk 2]
```diff
@@ -202,18 +203,25 @@ mod tests {
...
-        );
+        process
+            .chiplets
+            .write_mem(
+                ctx,
+                inputs[1].as_int().try_into().expect("Shouldn't fail by construction"),
+                a,
+            )
+            .unwrap();
+        process.execute_op(Operation::Noop).unwrap();
```
Comment (Al-Kindi-0) [inline processor/src/operations/comb_ops.rs:224]: Q: Is this related to your latest comment?

**Review by Al-Kindi-0 — APPROVED** [review]
Looks great, thank you!

File: processor/src/operations/comb_ops.rs (hunk 2 above)
Comment (plafer) [reply processor/src/operations/comb_ops.rs:224]: Yup!

File: processor/src/chiplets/memory/segment.rs [hunk 3]
```diff
@@ -72,37 +76,60 @@ impl MemorySegmentTrace {
...
     /// Writes the provided word at the specified address. The memory access is assumed to happen
     /// at the provided clock cycle.
-    pub fn write(&mut self, addr: u32, clk: Felt, value: Word) {
+    pub fn write(
+        &mut self,
+        ctx: ContextId,
+        addr: u32,
+        clk: Felt,
+        value: Word,
+    ) -> Result<(), ExecutionError> {
```
Comment (bobbinth) [inline processor/src/chiplets/memory/segment.rs:114]: Similar to the last comment, I'd update doc comments to explain under what conditions an error could happen.

File: processor/src/chiplets/memory/mod.rs [hunk 4]
```diff
@@ -137,15 +137,26 @@ impl Memory {
     ///
     /// If the specified address hasn't been previously written to, four ZERO elements are
     /// returned. This effectively implies that memory is initialized to ZERO.
-    pub fn read(&mut self, ctx: ContextId, addr: u32, clk: RowIndex) -> Word {
+    pub fn read(
```
Comment (bobbinth) [inline processor/src/chiplets/memory/mod.rs:140]: Same as above.

File: processor/src/chiplets/memory/mod.rs [hunk 5]
```diff
@@ -137,15 +137,26 @@ impl Memory {
...
 
     /// Writes the provided word at the specified context/address.
-    pub fn write(&mut self, ctx: ContextId, addr: u32, clk: RowIndex, value: Word) {
+    pub fn write(
+        &mut self,
+        ctx: ContextId,
+        addr: u32,
+        clk: RowIndex,
+        value: Word,
+    ) -> Result<(), ExecutionError> {
```
Comment (bobbinth) [inline processor/src/chiplets/memory/mod.rs:157]: Same comment as above.

File: miden/tests/integration/operations/io_ops/mem_ops.rs [hunk 6]
```diff
@@ -228,3 +231,23 @@ fn read_after_write() {
...
+#[test]
+fn mem_reads_same_clock_cycle() {
```
Comment (bobbinth) [inline miden/tests/integration/operations/io_ops/mem_ops.rs:241]: Should the doc comment here be updated?

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a few doc-related comments inline. Once these are addressed, we can merge.

File: processor/src/chiplets/memory/segment.rs [hunk 7]
```diff
@@ -72,37 +76,60 @@ impl MemorySegmentTrace {
     ///
     /// If the specified address hasn't been previously written to, four ZERO elements are
     /// returned. This effectively implies that memory is initialized to ZERO.
-    pub fn read(&mut self, addr: u32, clk: Felt) -> Word {
+    pub fn read(&mut self, ctx: ContextId, addr: u32, clk: Felt) -> Result<Word, ExecutionError> {
```
[the remark this answers is not in the document]
Comment (plafer) [reply processor/src/chiplets/memory/segment.rs:79]: > Also, the ctx is passed into here purely for error context tracking, right? That's probably fine for now, but I wonder if we should try to fix this in the future.

👍

**Pull request merged**